aiidalab_widgets_base/structures.py

```diff
@@ -4,126 +4,113 @@
 import tempfile
 import datetime
 from collections import OrderedDict
-from traitlets import Bool
+import numpy as np
 import ipywidgets as ipw
+from traitlets import Instance, Unicode, Union, link, observe, validate
 
-from aiida.orm import CalcFunctionNode, CalcJobNode, Node, QueryBuilder, WorkChainNode, StructureData
+# ASE imports
+from ase import Atom, Atoms
+from ase.data import chemical_symbols
+
+# AiiDA and AiiDA lab imports
+from aiida.orm import CalcFunctionNode, CalcJobNode, Data, QueryBuilder, Node, StructureData, WorkChainNode
+from aiida.plugins import DataFactory
 from .utils import get_ase_from_file
 
 
-class StructureManagerWidget(ipw.VBox):  # pylint: disable=too-many-instance-attributes
+class StructureManagerWidget(ipw.VBox):
     '''Upload a structure and store it in AiiDA database.
 
-    Useful class members:
-    :ivar has_structure: whether the widget contains a structure
-    :vartype has_structure: bool
-    :ivar frozen: whenter the widget is frozen (can't be modified) or not
-    :vartype frozen: bool
-    :ivar structure_node: link to AiiDA structure object
-    :vartype structure_node: StructureData or CifData'''
+    Attributes:
+        structure(Atoms): trait that contains the selected structure. 'None' if no structure is selected.
+        structure_node(StructureData, CifData): trait that contains AiiDA structure object
+        node_class(str): trait that contains structure_node type (as string).
+    '''
+
+    structure = Union([Instance(Atoms), Instance(Data)], allow_none=True)
```

Comment:
If I understand this correctly, the intention here is to be able to connect this traitlet to either an `Atoms` instance or a `StructureData` instance, correct? Why not be consistent, require one of the two types and then let the user use `dlinks` instead of `link` to connect to this traitlet. The two data types are very similar and by allowing the traitlet to be of either type, it is going to be very difficult to use it in practice.

Reply:
Please have a look into c91d1d3, not sure if it is an elegant solution, though. It turned out that 

1. ase1 and ase2 are not always the same after the following conversion:
ase1 -> Cif/Structure -> ase2

1. Cif/Structure are **never** the same after convering into `Atoms` and back.

So, in the end, I am not totally convinced that c91d1d3 is a good thing to do.

Reply:
This was discussed offline.